--- a/lib/iov_iter.c
+++ b/lib/iov_iter.c
@@ -1822,38 +1836,24 @@
 }
 EXPORT_SYMBOL(import_single_range);
 
-/**
- * iov_iter_restore() - Restore a &struct iov_iter to the same state as when
- *     iov_iter_save_state() was called.
- *
- * @i: &struct iov_iter to restore
- * @state: state to restore from
- *
- * Used after iov_iter_save_state() to bring restore @i, if operations may
- * have advanced it.
- *
- * Note: only works on ITER_IOVEC, ITER_BVEC, and ITER_KVEC
- */
-void iov_iter_restore(struct iov_iter *i, struct iov_iter_state *state)
-{
-	if (WARN_ON_ONCE(!iov_iter_is_bvec(i) && !iter_is_iovec(i)) &&
-			 !iov_iter_is_kvec(i))
-		return;
-	i->iov_offset = state->iov_offset;
-	i->count = state->count;
-	/*
-	 * For the *vec iters, nr_segs + iov is constant - if we increment
-	 * the vec, then we also decrement the nr_segs count. Hence we don't
-	 * need to track both of these, just one is enough and we can deduct
-	 * the other from that. ITER_KVEC and ITER_IOVEC are the same struct
-	 * size, so we can just increment the iov pointer as they are unionzed.
-	 * ITER_BVEC _may_ be the same size on some archs, but on others it is
-	 * not. Be safe and handle it separately.
-	 */
-	BUILD_BUG_ON(sizeof(struct iovec) != sizeof(struct kvec));
-	if (iov_iter_is_bvec(i))
-		i->bvec -= state->nr_segs - i->nr_segs;
-	else
-		i->iov -= state->nr_segs - i->nr_segs;
-	i->nr_segs = state->nr_segs;
-}
+int iov_iter_for_each_range(struct iov_iter *i, size_t bytes,
+			    int (*f)(struct kvec *vec, void *context),
+			    void *context)
+{
+	struct kvec w;
+	int err = -EINVAL;
+	if (!bytes)
+		return 0;
+
+	iterate_all_kinds(i, bytes, v, -EINVAL, ({
+		w.iov_base = kmap(v.bv_page) + v.bv_offset;
+		w.iov_len = v.bv_len;
+		err = f(&w, context);
+		kunmap(v.bv_page);
+		err;}), ({
+		w = v;
+		err = f(&w, context);})
+	)
+	return err;
+}
+EXPORT_SYMBOL(iov_iter_for_each_range);
